### ISSUE TITLE
os-bind - Thoughts Features

### ISSUE THREAD

**Issue by rhysxevans** [opnsense/plugins#974]
Hi

Firstly, I would like to say thank you, for your work on this, it is appreciated. However I would like to see the following added to the os-bind plugin

1) Host overides - basically move away from the Unbound package and solution "chaining"
2) Ability to enable safe search features (google, bing, youtube etc) from a check box type setup (I am currently doing this via unbound and the custom options) ( https://www.cwssoft.com/?p=1577 , https://forum.netgate.com/topic/105598/using-bind-to-enforce-google-safesearch/4)
3) Ability to identify blocking lists - Ie www.xxx.com was blocked by top porn list (bad example I know) 
4) MalwarePatrol list integration - this requires people to sign up, and each user is assigned a unique reciept key which will need to be factored in (https://www.malwarepatrol.net/tech-support/ & https://www.malwarepatrol.net/non-commercial/). I am particulary interested in the ransomware list, followed by the URL block Lists (they are quite agreesive and would require whitelist management)
5) Ability to do RPZ lists via a master slave bind DNS setup (OPNsense being the slave) (https://www.deteque.com/app/uploads/2018/04/Spamhaus-DNS-RPZ-DROP-Setup-v.2.pdf & https://deepthought.isc.org/article/AA-00518/0/How-can-I-synchronize-DNS-RPZ-firewall-policies-across-multiple-DNS-servers.html  & https://www.securityzones.net/images/downloads/BIND_RPZ_Installation_Guide.pdf)

I have previously done most of the above on linux variants before, however this has always been via CLI, and I wouldn't have the foggiest on how to present this via the GUI. So basically I am willing to help on the backend side, where/if required

Thanks

**Comment by mimugmail** [comment]
I'm planning full authoritative DNS for next year, but for interim, Overrides should be doable. Force safesearch should be easy, will add it.3. is not possible since all lists will be put in one file and kill duplicates. Malwarepatrol is a good idea, I have an account somewhere .. will have a look. :)

**Comment by rhysxevans** [comment]
Thanks, let me know, if you would like me to help out or test anything

**Comment by mimugmail** [comment]
With a separate RPZ zone, you mean just a new tab with key => value putting this to a zone like within telegraf:

![image](https://user-images.githubusercontent.com/20438689/48539081-7508c100-e8b6-11e8-91a6-ce05d6aac654.png)

**Comment by rhysxevans** [comment]
Hi

I may be being dumb (more than likely) , which feature above is this related to ? 

I thought you implied everything went into a single RPZ zone for the blacklists , multiple zones would allow for list differentiation of source logs in the logs ?

If it is regarding the walled garden question, that could be dealt with just by setting the "default" policy response to a CNAME that can be entered via the gui

Thanks

**Comment by mimugmail** [comment]
No .. not a single file, only the blacklists. I'm hoping you can read named.conf, perhaps you'll then understand my question:

```
acl "lan" { 192.168.0.0/24; 127.0.0.0/8; };

options {

        directory       "/usr/local/etc/namedb/working";
        pid-file        "/var/run/named/pid";
        dump-file       "/var/dump/named_dump.db";
        statistics-file "/var/stats/named.stats";

         listen-on port 53530 { 127.0.0.1; };
         listen-on-v6 port 53530 { ::1; };


        response-policy { zone "whitelist.localdomain"; zone "blacklist.localdomain";zone "rpzgoogle";zone "rpzyoutube"; };

        recursion          yes;
        allow-recursion    { lan; };


        max-cache-size    80%;

        dnssec-validation    no;

};

key "rndc-key" {
        algorithm hmac-sha256;
        secret "VxtIzJevSQXqnr7h2qerrcwjnZlMWSGGFBndKeNIDfw=";
};
controls {
        inet 127.0.0.1 port 9530
                allow { 127.0.0.1; } keys { "rndc-key"; };
};

zone "." { type hint; file "/usr/local/etc/namedb/named.root"; };

zone "localhost"        { type master; file "/usr/local/etc/namedb/master/localhost-forward.db"; };
zone "127.in-addr.arpa" { type master; file "/usr/local/etc/namedb/master/localhost-reverse.db"; };
zone "0.ip6.arpa"       { type master; file "/usr/local/etc/namedb/master/localhost-reverse.db"; };

zone "whitelist.localdomain" { type master; file "/usr/local/etc/namedb/master/whitelist.db"; notify no; check-names ignore; };
zone "blacklist.localdomain" { type master; file "/usr/local/etc/namedb/master/blacklist.db"; notify no; check-names ignore; };
zone "rpzgoogle" { type master; file "/usr/local/etc/namedb/master/google.db"; notify no; check-names ignore; };
zone "rpzyoutube" { type master; file "/usr/local/etc/namedb/master/youtube.db"; notify no; check-names ignore; };

logging {
        channel default_log {
                file "/var/log/named/named.log" versions 3 size 7m;
                print-time yes;
                print-severity yes;
                print-category yes;
        };

        channel query_log {
                file "/var/log/named/query.log" versions 3 size 7m;
                print-time yes;
        };

        channel rpz_log {
                file "/var/log/named/rpz.log" versions 3 size 7m;
                print-time yes;
        };

        category default { default_log; };
        category general { default_log; };
        category queries { query_log; };
        category rpz { rpz_log; };
};
```

**Comment by rhysxevans** [comment]
Hi

Ok, so your talking about the safe search stuff the "zone "rpzgoogle"" and "zone "rpzyoutube"" in this case ?

If so the proposed above would work.

I assume Services -> Bind -> Configuration -> SafeSearch Tab -> tick boxes ?

Thanks

**Comment by mimugmail** [comment]
No, I mean the walled garden thing. Can you add a sample named.conf and a zone how you would use a walled garden. Perhaps I'm intepreting walled garden in a different way than you do :)

**Comment by mimugmail** [comment]
OK, so how you describe it lokal Overrides and walled Garden could be in the same file

**Comment by rhysxevans** [comment]
Hi

I would presume they would be different, however a walled garden may require a host override for the CNAME entry. 

1) Basically in RPZ blacklist configuration, setup required redirect CNAME 
2) Setup a Host Override for the CNAME created above (this could be on public name servers aswell)

Hopefully I am making sense

Thanks

**Comment by rhysxevans** [comment]
Hi

Just confirmed the syntax, slight typo in the above example (responsepolicy should be response-policy)

So

```
response-policy { 
		zone "white" policy PASSTHRU;
		zone "black" policy CNAME walled.garden;
}
```
But essentially the format is correct, and walled.garden could be any valid DNS entry

Thanks

**Comment by mimugmail** [comment]
Ok, I belive I know how to implement it. It should be finished next week. Have to check some constraint options with @fichtner first :)

**Comment by rhysxevans** [comment]
great, thanks, let me know if you need me to do anything

**Comment by rhysxevans** [comment]
Hi, 

Just updated and have flipped over to BIND for SafeSearch , working like a charm.

Thank you

**Comment by mimugmail** [comment]
Next update brings also safe search for Bind, then I'll start with Overrides ...

**Comment by mimugmail** [comment]
@stilez can you post your exact config please? Is this some RPZ related rewriting what you need?

**Comment by stilez** [comment]
> @stilez can you post your exact config please? Is this some RPZ related rewriting what you need?

I'm still working it out, it's going to take a while to get good. Tentatively, it _might_ look like this:

```
options {
        directory       "/usr/local/etc/namedb/working";
        pid-file        "/var/run/named/pid";
        dump-file       "/var/dump/named_dump.db";
        statistics-file "/var/stats/named.stats";

	// service hardening

	version none;
	hostname none;
	rate-limit {
                responses-per-second 10;
                exempt-clients { 127.0.0.1; ::1; };
        };

	// authoritative server only

        recursion no;
        fetch-glue no;
        allow-transfer { none; };
        allow-query { localhost;LETSENCRYPT'S_IP_OR_SUBNET_FOR_QUERYING; };
	// general service definitions
        listen-on port ******* { localhost;LETSENCRYPT'S_IP_OR_SUBNET_FOR_QUERYING; };
        dnssec-validation    auto;  // TO SET UP LATER
};

logging {
        [ default OPNSense config]
};

// "." ROOT ZONE HINT - should be OMITTED/COMMENTED OUT for authoritative server

zone "acme-validation.example.com" {
	type master;
	file "/etc/bind/zones/db.acme-validation.example.com";
	// allow-transfer { IP_ADDR_TO_ADD ; };	 // ENABLE WHEN SLAVE NS EXISTS
        notify yes;
        [ OTHER ZONE DATA / FILE LINK]
};
```

My scenario:  I want to use Let's Encrypt certbot DNS-01, but certbot can't use my ISP's DNS API. The domains are private use only - I only need DNS for myself, when I'm accessing them remotely via HTTPS (and need the TLS cert chain to work from other locations), nobody else uses them. Certbot *can* be used via RFC2136 with Bind though, so if I create a "dummy/fake" domain (`acme-validation.mydomain.com`), redireect the usual validation domain from my ISP to Bind's public IP, and use Bind locally as the authoritative NS for the acme validation domain, I can get acme/certrbot working nicely.  

The only time Bind should get queried is once every 60 - 90 days, when Let's Encrypt wants to check the TXT records for `acme-validation.mydomain.com`. There are no other records, and the domain doesn't actually exist anywhere outside Bind.

Because that's a very narrow use-case, I can lock down Bind completely. I am using Unbound for my local resolver and want to continue doing so, listening on the LAN interface only. I enable Bind on port 53 of an unused public IP on the WAN interface with no recursion/forwarding, no root zone hint, heavy rate limiting, and one zone defined only, a bit like above. And with luck - acme/Let's Encrypt will now work, even though my ISP's DNS API isn't compatible. But it's a very customised config, not a common use-case (or is it?) so I can't see myself doing this purely using the OPNSense GUI.  I need to be able to manually create named.conf content lines.

**Comment by mimugmail** [comment]
Because your setup is so specific it should be easy to implement.
What is the reason for excluding hint file? As you disallow recursion and only allow your provider to query?

And what are the contents of your zonefile?

**Comment by stilez** [comment]
> What is the reason for excluding hint file? As you disallow recursion and only allow your provider to query?
> And what are the contents of your zonefile?

The hint is disabled based on [this ISC technical note](https://ftp.isc.org/isc/pubs/tn/isc-tn-2002-2.html), which says to separate caching and authoritative nameservers, and explains the implication of a root hint for them. As no other NS will be updated in a timely manner, or at all, with the ACME TXT validation data, there's no point in providing a hint to another NS, and it's detrimental in the context - _at best_ it will cause Let's Encrypt's system to look at other NS outdated versions of the TXT data if for any reason it doesn't reach Bind, rather than note the failure or try again, causing an explicit validation error (wrong TXT data found) rather than a retry or "can't connect" error, and may cache the bad result as well.

I don't yet know the zonefile contents I will need - I'm figuring this out (new to Bind, only using it because of ISP/provider API limitations, otherwise I'd do it via ISP/provider and not need Bind at all).

**Comment by mimugmail** [comment]
Ok, got it. Let's do it the other way around. What exactly to you need drom LE DNS-01 to work. Is there a technical doc from LE which explains for every DNS software how to configure? Maybe this way I can build a workaround for you.

My LE knowledge is sadly near 0 :)

**Comment by stilez** [comment]
To avoid jamming this thread with a slight side topic, let's discuss LE and its DNS-01 system by email, so we can focus here, on Bind plugin features/enhancements.  We can post back here, any conclusions or thoughts.  I've sent you a PM on the OPNsense forum with my email.

**Comment by apiods** [comment]
Thanks for the work on the BIND package - I have it working nicely also :-)

Just wanted to add support for a couple of the feature requests:

- DNSBL local Blacklist domains (opposite of the whitest feature)
- Advanced config option: support forwarder/secondary zones

Thanks.

**Comment by macktic** [comment]
Thanks for all the hard work. 
I’ve read all above, but would still request a custom option somehow. 
Even if it would be in a file on the file system and only available via ssh.  I’m currently adding options that way, but they get deleted on save or update. 

Anyway, the option I’m after is to strip all IPv6 answers ( filter-aaaa-on-v4 yes; )
I don’t need bind to listen on an IPv6 address either, but that’s currently not possible in the gui either, as I’ll get an error if I don’t put anything in the listen IPv6 address. 

Some background:
My ISP doesn’t support IPv6 at the moment, so I’ve got IPv6 functionality via HE net tunnelbroker. 

The issue is that Netflix detects this as a possible way to get around their blocks and stops working. 

The easiest way round this is by stripping all IPv6 answers from any of the Netflix servers. 

I do this (as I used to on pfsense) by using unbound as normal dns server and use host overrides to send the Netflix servers to bind (running on a non standard port) in this scenario I don’t need bind to give me any IPv6 answers ever.

**Comment by mimugmail** [comment]
So, all you need is to make v6 field optional and a checkbox to strip v6 answers? 
Thats easy :)

Next version will also offer zone management

**Comment by L1ghtn1ng** [comment]
@mimugmail Would be good if we could have requests go over DoH using cloudflare https://developers.cloudflare.com/1.1.1.1/dns-over-https/

**Comment by mimugmail** [comment]
Does bind support DoH yet? Perhaps we need a stunnel plugin ;) https://kb.isc.org/docs/aa-01386

**Comment by L1ghtn1ng** [comment]
Not that I am aware of but using a forwarder that would somehow point
to  cloudflare would work right? /home/pi/argo-tunnel/cloudflared
proxy-dns --port 54 --upstream https://1.1.1.1/.well-known/dns-query --
upstream https://1.0.0.1/.well-known/dns-query --upstream 
https://2606:4700:4700::1111/.well-known/dns-query --ups

Having to use this on my pihole currently to have the DoH support and
just modified the config to point to that

Just would be nice to have a solution with no workarounds

**Comment by mimugmail** [comment]
Why not using dnscrypt proxy plugin? I'm using it as main DNS on Firewall, also with DNSBL support.

**Comment by L1ghtn1ng** [comment]
Can you create a how to for it please?
________________________________

**Comment by mimugmail** [comment]
Next time you try it on your own ;)
https://github.com/opnsense/docs/pull/169

**Comment by L1ghtn1ng** [comment]
Thank you 🤗
________________________________

**Comment by mimugmail** [comment]
1, 2, 3, 5 done, close?

**Comment by fichtner** [comment]
Sure, 4 can be re-requested.

**Comment by zanechua** [comment]
@mimugmail was the option for ipv6 filtering ever added in?

For `filter-aaaa-on-v4`? I don't see it in the UI for opnsense for the most part

**Comment by mimugmail** [comment]
Not that I'm aware of, how is the exact Syntax?

**Comment by zanechua** [comment]
It just needs to be enabled during the binary build with `./configure --enable-filter-aaaa`

And usage is just simply
```
filter-aaaa-on-v4 yes;
filter-aaaa-on-v6 yes;
```

Docs: https://kb.isc.org/docs/aa-00576

**Comment by zanechua** [comment]
@mimugmail 

Actually. Looking at the Makefile, the os-bind plugin is using bind914. 

bind916 is the next ESV version after bind911. Curious why bind 911 isn't being used since it's EOL in 2021 Q4 while bind914 is EOL 2020 Q1 and bind916 is EOL 2024 Q1.

Any issues reverting to bind911 since that will still be supported? bind911 has the filter-aaaa as default build options.

I've already submitted a pull to add it into bind916 upstream.

Bind support timeline: https://kb.isc.org/docs/aa-00896

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=244332
